core/embed/rust/src/ui/model_tt/component/keyboard/word_count.rs

```diff
@@ -8,33 +8,99 @@ use crate::ui::{
     shape::Renderer,
 };
 
-const NUMBERS: [u32; 5] = [12, 18, 20, 24, 33];
-const LABELS: [&str; 5] = ["12", "18", "20", "24", "33"];
-const CELLS: [(usize, usize); 5] = [(0, 0), (0, 2), (0, 4), (1, 0), (1, 2)];
-
-pub struct SelectWordCount {
-    button: [Button; NUMBERS.len()],
-}
-
 pub enum SelectWordCountMsg {
     Selected(u32),
 }
 
+// We allow large_enum_variant here because the code is simpler and the larger
+// variant (ValueKeypad<5>) predates the smaller one.
+#[allow(clippy::large_enum_variant)]
+pub enum SelectWordCount {
+    All(ValueKeypad<5>),
+    Multishare(ValueKeypad<2>),
+}
+
 impl SelectWordCount {
-    pub fn new() -> Self {
-        SelectWordCount {
-            button: LABELS.map(|t| Button::with_text(t.into()).styled(theme::button_pin())),
-        }
+    const NUMBERS_ALL: [u32; 5] = [12, 18, 20, 24, 33];
+    const LABELS_ALL: [&'static str; 5] = ["12", "18", "20", "24", "33"];
+    const CELLS_ALL: [(usize, usize); 5] = [(0, 0), (0, 2), (0, 4), (1, 0), (1, 2)];
+
+    const NUMBERS_MULTISHARE: [u32; 2] = [20, 33];
+    const LABELS_MULTISHARE: [&'static str; 2] = ["20", "33"];
+    const CELLS_MULTISHARE: [(usize, usize); 2] = [(0, 0), (0, 2)];
+
+    pub fn new_all() -> Self {
+        Self::All(ValueKeypad::new(
+            Self::NUMBERS_ALL,
+            Self::LABELS_ALL,
+            Self::CELLS_ALL,
+        ))
+    }
+
+    pub fn new_multishare() -> Self {
+        Self::Multishare(ValueKeypad::new(
+            Self::NUMBERS_MULTISHARE,
+            Self::LABELS_MULTISHARE,
+            Self::CELLS_MULTISHARE,
+        ))
     }
 }
 
 impl Component for SelectWordCount {
     type Msg = SelectWordCountMsg;
 
+    fn place(&mut self, bounds: Rect) -> Rect {
+        match self {
+            SelectWordCount::All(full_selector) => full_selector.place(bounds),
+            SelectWordCount::Multishare(bin_selector) => bin_selector.place(bounds),
+        }
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        match self {
+            SelectWordCount::All(full_selector) => full_selector.event(ctx, event),
+            SelectWordCount::Multishare(bin_selector) => bin_selector.event(ctx, event),
+        }
+    }
+
+    fn paint(&mut self) {
+        match self {
+            SelectWordCount::All(full_selector) => full_selector.paint(),
+            SelectWordCount::Multishare(bin_selector) => bin_selector.paint(),
+        }
+    }
+
+    fn render<'s>(&'s self, target: &mut impl Renderer<'s>) {
+        match self {
+            SelectWordCount::All(full_selector) => full_selector.render(target),
+            SelectWordCount::Multishare(bin_selector) => bin_selector.render(target),
+        }
+    }
+}
+
+pub struct ValueKeypad<const N: usize> {
```

Comment:
This is still the case where rustc makes copies of methods for each generic instance. Using Vec or array slices would result in 2x less code generated.

Reply:
I considered that the generic parameter makes it duplicated but:
- the Component is very small
- this is for Model T only, where the space is not a problem. 

Anyway, I tried using `heapless::Vec`, I tried in two arrangements:
1. `Vec` for each of these, like this
```rust
pub struct ValueKeypad {
    buttons: Vec<Button, 5>,
    numbers: Vec<u32, 5>,
    cells: Vec<(usize, usize), 5>,
}
```
2. `Vec` for a tuple of everything, something like this:
```rust
type ValueKeyPacked = (Button, u32, (usize, usize));
pub struct ValueKeypad {
    buttons: Vec<ValueKeyPacked, 5>,
}
```
Both versions result in the same binsize. Do you have other suggestions?

Reply:
2. seems nice ... though I didn't realize the memory situation on T2T1 is less worse so feel free to leave it as it is with a comment warning against adding further instances.

Reply:
So I measured again and `Vec` actually saves one block of flash on debug build. Not much but we might prevent bloat by other variants potentially in the future. So I implemented the version with Vec.

@mmilata ftr, the situation on `T2T1` with this change
`FLASH` - 83.72% used
`FLASH2:` - 92.75% used
d662aa1c2307182080af4ee375d7b7bed60d5bcf